File: algorithm/src/array/removeDuplicates.c
#include "alg_array.h"

int removeDuplicates( int* nums, int numsSize ){
    /*
     * i 遍历数组下标
     * j 删除重复后的数组下标
     */
    int i, j;

    /* IN 0 OUT 0
     * IN 1 OUT 1
     */
    if( nums == NULL || 2 > numsSize ){
        return numsSize;
    }

    for( i = 1,j = 0; i < numsSize; i++ ){
       if( nums[i] == nums[j] ){
           continue; 
       }
       j++;
       if( j != i ){ //优化条件 避免原地复制
           nums[j]=nums[i];
       }
    }
    return j+1;
}

File: algorithm/src/array/removeDuplicates.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

class Solution:
    def removeDuplicates(self, nums):
        if( len(nums) < 2 ):
            return len(nums);
        j = 1;
        for i in range(1, len(nums)):
            print("i,nums[i],j,nums[j],nums[j-1]",i,nums[i],j,nums[j],nums[j-1])
            if nums[i] != nums[j-1]:
                nums[j] = nums[i];
                j += 1
        return j;

def main():
    l = [0, 0, 0, 1, 1, 2, 2, 3, 3, 3, 4, 4, 5];
    s = Solution();
    print( s.removeDuplicates(l));
    print(l);

if __name__ == '__main__':
    main();
